File: 20/20.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        mapping = {')': '(', '}': '{', ']': '['}

        for char in s:
            if char in mapping.values():
                stack.append(char)
            elif char in mapping:
                if not stack:
                    return False
                top_element = stack.pop()
                if mapping[char] != top_element:
                    return False
            else:
                return False

        return not stack

solution = Solution()
print(solution.isValid("()"))
print(solution.isValid("()[]{}"))
print(solution.isValid("(]"))

File: 20/20.rb
def is_valid(s)
    stack = []
    mapping = {')' => '(', '}' => '{', ']' => '['}

    s.each_char do |char|
        if mapping.value?(char)
            stack.push(char)
        elsif mapping.key?(char)
            top_element = stack.empty? ? '#' : stack.pop
            if mapping[char] != top_element
                return false
            end
        else
            return false
        end
    end

    stack.empty?
end

puts is_valid("()")
puts is_valid("()[]{}")
puts is_valid("(]")
